pkg/kp/pcstore/consul_store.go

```diff
@@ -308,6 +309,147 @@ func (s *consulStore) WatchPodCluster(id fields.ID, quit <-chan struct{}) <-chan
+// are paired together in a map of pc ID -> change objects. Each change will be sent
+// to the respective sync channels of each pod cluster later on.
+func (s *consulStore) zipResults(current, previous WatchedPodClusters) map[fields.ID]podClusterChange {
+	allPrevious := make(map[fields.ID]*fields.PodCluster)
+	for _, prev := range previous.Clusters {
+		allPrevious[prev.ID] = prev
+	}
+	ret := map[fields.ID]podClusterChange{}
+	for _, cur := range current.Clusters {
+		prev, ok := allPrevious[cur.ID]
+		ret[cur.ID] = podClusterChange{
+			previous: prev,
+			current:  cur,
+		}
+		if ok {
+			delete(allPrevious, cur.ID)
+		}
+	}
+	for _, prev := range allPrevious {
+		ret[prev.ID] = podClusterChange{
+			previous: prev,
+		}
+	}
+	return ret
+}
+
+// try forever to match the expectations as defined in the provided change channel.
+// If a change fails to take, this function will retry that change forever until
+// it works as expected or a newer change appears on the channel. This routine also
+// executes and monitors the label watch for the pod's label selector.
+// TODO: since most of this code is generic to pod cluster syncers, this will likely
+// be extracted to github.com/square/p2.
+func (s *consulStore) handlePCUpdates(concrete ConcreteSyncer, changes chan podClusterChange) {
+	var change podClusterChange
+	podWatch := make(chan []labels.Labeled)
+	podWatchQuit := make(chan struct{})
+	defer func() {
+		close(podWatchQuit)
+	}()
+
+	for {
+		var ok bool
+
+		select {
+		case labeledPods := <-podWatch:
+			s.logger.Debugf("Calling SyncCluster with %v / %v", change.current, labeledPods)
+			concrete.SyncCluster(change.current, labeledPods)
+		case change, ok = <-changes:
+			if !ok {
+				return // we're closed for business
+			}
+
+			if change.current == nil && change.previous != nil {
+				// if no current cluster exists, but there is a previous cluster,
+				// it means we need to destroy this concrete cluster
+				s.logger.WithField("pc_id", change.previous.ID).Infof("Calling DeleteCluster with %v", change.previous)
+				err := concrete.DeleteCluster(change.previous)
+				if err != nil {
+					s.logger.Errorf("Deletion of cluster failed! %v", err)
+				} else {
+					return
+				}
+			} else if change.current != nil && change.previous != nil {
+				// if there's a current and a previous pod cluster, update concrete cluster metadata and
+				// refresh the pod selector watch if it changed
+				if change.current.PodSelector.String() != change.previous.PodSelector.String() {
+					close(podWatchQuit)
+					podWatchQuit = make(chan struct{})
+					s.logger.WithFields(logrus.Fields{
+						"pc_id":        change.current.ID,
+						"old_selector": change.previous.PodSelector.String(),
+						"new_selector": change.current.PodSelector.String(),
+					}).Debugf("Altering pod selector for %v", change.current.ID)
+					podWatch = s.applicator.WatchMatches(change.current.PodSelector, labels.POD, podWatchQuit)
+				}
```

Comment:
You don't want to call `concrete.SyncCluster` here because you don't have any results from the `podWatch` yet. The new annotations will be provided when `concrete.SyncCluster` is called after the `podWatch` fires.

It does mean that there is a somewhat artificial delay in updating annotations since you have to wait on the label watcher. In practice, the Consul label watcher is guaranteed to fire every 10 seconds at least, so this is probably fine but not great.

I think maybe we could keep the last result from `podWatch` as a variable out of scope of the `for` loop and sync the cluster immediately to get the desired effect. Going to think about whether that's safe to do.